compiler/semantic/op.go

```diff
@@ -640,22 +640,27 @@ func (a *analyzer) semOp(o ast.Op, seq dag.Seq) (dag.Seq, error) {
 	case *ast.Rename:
 		var assignments []dag.Assignment
 		for _, fa := range o.Args {
-			dst, err := a.semField(fa.LHS)
+			dst, err := a.semExpr(fa.LHS)
 			if err != nil {
-				return nil, errors.New("rename: requires explicit field references")
+				return nil, fmt.Errorf("rename: %w", err)
 			}
-			src, err := a.semField(fa.RHS)
+			if !isLval(dst) {
+				return nil, fmt.Errorf("rename: illegal left-hand side of assignment")
+			}
+			src, err := a.semExpr(fa.RHS)
 			if err != nil {
-				return nil, errors.New("rename: requires explicit field references")
+				return nil, fmt.Errorf("rename: %w", err)
 			}
-			if len(dst.Path) != len(src.Path) {
-				return nil, fmt.Errorf("rename: cannot rename %s to %s", src, dst)
+			if !isLval(src) {
+				return nil, fmt.Errorf("rename: illegal right-hand side of assignment")
 			}
-			// Check that the prefixes match and, if not, report first place
-			// that they don't.
-			for i := 0; i <= len(src.Path)-2; i++ {
-				if src.Path[i] != dst.Path[i] {
-					return nil, fmt.Errorf("rename: cannot rename %s to %s (differ in %s vs %s)", src, dst, src.Path[i], dst.Path[i])
+			// If both paths are static validate them. Otherwise this will be
+			// done at runtime.
+			srcThis, srcOk := src.(*dag.This)
+			dstThis, dstOk := dst.(*dag.This)
+			if dstOk && srcOk {
+				if err := expr.CheckRenameField(srcThis.Path, dstThis.Path); err != nil {
+					return nil, fmt.Errorf("rename: %w", err)
 				}
 			}
 			assignments = append(assignments, dag.Assignment{Kind: "Assignment", LHS: dst, RHS: src})
```

Comment:
Simplify with semAssignments?
```go
			as, err := a.semAssignment(fa)
			if err != nil {
				return nil, fmt.Errorf("rename: %w", err)
			}
			if !isLval(as.RHS) {
				return nil, fmt.Errorf("rename: illegal right-hand side of assignment")
			}
			// If both paths are static validate them. Otherwise this will be
			// done at runtime.
			lhs, lhsOk := as.LHS.(*dag.This)
			rhs, rhsOk := as.RHS.(*dag.This)
			if lhsOk && rhsOk {
				if err := expr.CheckRenameField(lhs.Path, rhs.Path); err != nil {
					return nil, fmt.Errorf("rename: %w", err)
				}
			}
			assignments = append(assignments, as)
```

Reply:
I like it.